src/libcore/array.rs

```diff
@@ -285,3 +304,171 @@ macro_rules! array_impl_default {
+    /// Returns the remaining items of this iterator as a slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(array_into_iter)]
+    ///
+    /// # fn main() {
+    /// let array = ['a', 'b', 'c'];
+    /// let mut into_iter = array.into_iter();
+    /// assert_eq!(into_iter.as_slice(), &['a', 'b', 'c']);
+    /// let _ = into_iter.next().unwrap();
+    /// assert_eq!(into_iter.as_slice(), &['b', 'c']);
+    /// # }
+    /// ```
+    #[inline]
+    #[unstable(feature = "array_into_iter", issue = "0")]
+    pub fn as_slice(&self) -> &[T] {
+        &self.array.as_slice()[self.index..self.index_back]
+    }
+
+    /// Returns the remaining items of this iterator as a mutable slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(array_into_iter)]
+    ///
+    /// # fn main() {
+    /// let array = ['a', 'b', 'c'];
+    /// let mut into_iter = array.into_iter();
+    /// assert_eq!(into_iter.as_slice(), &['a', 'b', 'c']);
+    /// into_iter.as_mut_slice()[2] = 'z';
+    /// assert_eq!(into_iter.next().unwrap(), 'a');
+    /// assert_eq!(into_iter.next().unwrap(), 'b');
+    /// assert_eq!(into_iter.next().unwrap(), 'z');
+    /// # }
+    /// ```
+    #[inline]
+    #[unstable(feature = "array_into_iter", issue = "0")]
+    pub fn as_mut_slice(&mut self) -> &mut [T] {
+        &mut self.array.as_mut_slice()[self.index..self.index_back]
+    }
+}
+
+#[unstable(feature = "array_into_iter", issue = "0")]
+impl<T, A: FixedSizeArray<T>> Drop for IntoIter<T, A> {
+    #[inline]
+    fn drop(&mut self) {
+        // Drop values that are still alive.
+        for p in self.as_mut_slice() {
+            unsafe { ptr::drop_in_place(p); }
+        }
+    }
+}
+
+#[unstable(feature = "array_into_iter", issue = "0")]
+impl<T: Clone, A: FixedSizeArray<T>> Clone for IntoIter<T, A> {
+    fn clone(&self) -> Self {
+        unsafe {
+            let mut iter = Self {
+                array: ManuallyDrop::new(mem::uninitialized()),
+                index: 0,
+                index_back: 0,
+                _marker: PhantomData,
+            };
+
+            // Clone values that are still alive.
+            for (dst, src) in iter.array.as_mut_slice().iter_mut().zip(self.as_slice()) {
+                ptr::write(dst, src.clone());
+                iter.index_back += 1;
+            }
+
+            iter
+        }
+    }
+}
+
+#[unstable(feature = "array_into_iter", issue = "0")]
+impl<T: fmt::Debug, A: FixedSizeArray<T>> fmt::Debug for IntoIter<T, A> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("IntoIter")
+            .field(&self.as_slice())
+            .finish()
+    }
+}
+
+#[unstable(feature = "array_into_iter", issue = "0")]
+impl<T, A: FixedSizeArray<T>> Iterator for IntoIter<T, A> {
+    type Item = T;
+
+    #[inline]
+    fn next(&mut self) -> Option<T> {
+        if self.index < self.index_back {
+            let p = &self.array.as_slice()[self.index];
```

Comment:
Should these indexes use `ptr::offset` so as not to acquire `&[T]` to a partially uninitialized `[T]`?

The same applies down in `next_back`.

Reply:
Is an uninitialized slice really a problem if we never read from those parts?  I would think it's fine, otherwise sequences like `mem::uninitialized()` + `ptr::write()` would be a problem too.

Reply:
It's certainly problematic under the [types-as-contracts](https://www.ralfj.de/blog/2017/08/11/types-as-contracts-evaluation.html) view. AFAICT the unsafe code guidelines aren't solid enough to really give a hard answer yet, but it would certainly be better to not create invalid slices. Incidentially, yes, `mem::uninitialized` is problematic :)

Reply:
If that's the case, how can I even get the pointer base to offset from?  `Deref for ManuallyDrop` will create a momentary reference here too.  Or even with a local union type, getting the pointer to the field will require forming a reference first.

Reply:
I'd add an unstable `ManuallyDrop::as_ptr`.

Reply:
At the moment, `repr(transparent)` is only allowed on a `struct`.  The RFC left unions as future work: https://github.com/rust-lang/rfcs/pull/1758#issuecomment-290757356

Reply:
I don't think you need repr(transparent) for this, since this is just a question of memory layout, not calling convention. repr(C) would suffice, or inside libstd we could even simply rely on what rustc currently does (which should place the union contents at the same address as the union).

Reply:
@rkruppe I've been thinking more about this, and if my conclusion in https://github.com/rust-lang/rust/pull/49056#comment-373635380 is correct, then this PR would be fine with the current "invalid slice" usage, since the values left in the slice would still be valid values of type T, just ones in the "dropped" typestate.

Reply:
FWIW, here's a comparison using pointers to avoid possibly dropped/uninit refs:
https://github.com/cuviper/rust/compare/array-intoiter...cuviper:array-intoiter-ptr

I can submit the `ManuallyDrop` additions in a separate PR, if you think it's worth it.

> since the values left in the slice would still be valid values of type T, just ones in the "dropped" typestate.

The `Clone` implementation will create truly uninitialized entries too.  IMO the unsafe rules should not declare such references invalid if they're not read, but we can deal with it in pointers if necessary.

Reply:
Yes, that clone implementation is certainly become UB due to the uninitialized, but that's true of basically every use of mem::uninitialized, so we'll have to go clean them all up later no matter what once the uninit RFC lands